Authorized tenant and category filtering

koevskinikola opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
Filter audit logs by authorized tenants and categories.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #42789, #42736, #41120, #41231, #41211, [#42132](https://github.com/camunda/camunda/issues/42132)

koevskinikola committed:
feat: enable tenant filtering for audit log queries

Include authorized tenant filtering in the audit log
query in the search layer.

This commit also fixes the auditLogKey filter, as in
ES/OS the auditLogKey is the AuditLogEntity ID.

koevskinikola committed:
feat: add composite authorizations to audit log services

koevskinikola committed:
test: expand audit log acceptance tests

Split the audit log acceptance tests into
client and authorization IT tests. The goal
is to improve readability of the tests and make
their purpose clearer, so future test cases
can be added to the appropriate class.

eppdot commented on search/search-client-query-transformer/src/main/java/io/camunda/search/clients/transformers/filter/AuditLogFilterTransformer.java:
```diff
@@ -56,4 +56,28 @@ public SearchQuery toSearchQuery(final AuditLogFilter filter) {
+  }
+
+  @Override
+  protected SearchQuery toAuthorizationCheckSearchQuery(final Authorization<?> authorization) {
```
❌ it's not clear to me what and why this is done 🤔 which is a sign that this should be refactored to be more expressive or a least some explaining comments need to be added.

eppdot replied:
I think here it would help to explain via comments what this does and why it does it like this, e.g. why are the resourceIds mapped o the `CATEGORY` and also mapped to the `PROCESS_DEFINTION_ID` -> that just doesn't fit my thinking.

koevskinikola replied:
The authorization query creation is fixed now.

eppdot commented on search/search-domain/src/main/java/io/camunda/search/entities/AuditLogEntity.java:
```diff
@@ -43,6 +43,11 @@ public record AuditLogEntity(
     implements TenantOwnedEntity {
 
+  @Override
+  public boolean isInTenantScope() {
```
🔧 rename to `hasTenantScope`

eppdot commented on service/src/main/java/io/camunda/service/AuditLogServices.java:
```diff
@@ -72,8 +81,14 @@ public AuditLogEntity getAuditLog(final String auditLogKey) {
                         authentication,
-                        withAuthorization(
-                            AUDIT_LOG_READ_AUTHORIZATION, al -> al.category().name())))
+                        AuthorizationConditions.anyOf(
```
🔧 this is pretty much duplicated with line 55, should we store this in a constant/class? or extract into a method?

eppdot commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AuditLogIT.java:
```diff
@@ -143,6 +144,16 @@ static void setUp() {
     deployResource(adminClient, "processB.bpmn", PROCESS_B, TENANT_B);
+
+    final var useRest = true;
+    final var useGrpc = false;
```
💭 use parameterized tests here?

eppdot commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AuditLogIT.java:
```diff
@@ -153,149 +164,105 @@ void cleanUp() {
-    }
+  @Test
+  void shouldFetchAuditLogEntryByAuditLogKey() {
+    Awaitility.await("Audit log entry is created and can be fetched by audit log key")
```
💭 this is all prep work right? couldn't it even go into the before all?

koevskinikola replied:
I see your point. Instead of "awaiting" in every test, we can just do it in the `BeforeAll` and then skip this code in the test cases, as we're operating on the assumption that this was already confirmed in the `BeforeAll` method.

eppdot reviewed (changes requested):
Good Job Nikola.

A few things are not 100% clear to me (could also be because I am not super familiar with identity/auth checks), see the comments I left.

Also I think we could add some more non-IT tests especially for the authorization logic etc.

We can also jump on a sync call to accelerate 😄

eppdot commented on webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/template/AuditLogTemplate.java:
```diff
@@ -13,13 +13,15 @@
   public static final String INDEX_NAME = "audit-log";
   public static final String INDEX_VERSION = "8.9.0";
 
+  public static final String ID = "id";
```
Hmm, let's keep it simple. Sounds like it would complicate things more 😅

koevskinikola committed:
refactor: rename tenant scope method

Refactor the TenantOwned#isInTenantScope method
for better readability.

koevskinikola commented on service/src/main/java/io/camunda/service/AuditLogServices.java:
```diff
@@ -49,8 +52,14 @@ public SearchQueryResult<AuditLogEntity> search(final AuditLogQuery query) {
+                                PROCESS_INSTANCE_READ_AUTHORIZATION,
+                                al -> al.processDefinitionId()),
+                            withAuthorization(
+                                USER_TASK_READ_AUTHORIZATION, al -> al.processDefinitionId()))))
```
That's part of the `AuditLogFilterTransformer`. I've included an extra query for `USER_TASK` audit logs there.

eppdot commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AbstractAuditLogClientIT.java:
```diff
@@ -0,0 +1,152 @@
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.Test;
+
+abstract class AbstractAuditLogClientIT {
```
💭 Maybe a better name could be `BasicAuditLogClientIT` or `Base...` - it feels a bit strange that an `abstract` test has `@Test` case 🤔

eppdot commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AuditLogAuthorizationsIT.java:
```diff
@@ -0,0 +1,275 @@
+
+@MultiDbTest
+@DisabledIfSystemProperty(named = "test.integration.camunda.database.type", matches = "AWS_OS")
+public class AuditLogAuthorizationsIT {
```
🔧 the setup code is quite similar to the `Abstract...` test, but this one doesn't inherit from it.

Could we prefer composition over inheritance and just move things into a shared helper? Then we could get rid of `abstract` and the inheritances stuff.

eppdot commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AbstractAuditLogClientIT.java:
```diff
@@ -0,0 +1,152 @@
+        .isEqualTo(String.valueOf(processInstanceEvent.getProcessInstanceKey()));
+    assertThat(fetchedAuditLog.getTenantId()).isEqualTo(TENANT_A);
+    assertThat(fetchedAuditLog.getResult()).isEqualTo(AuditLogResultEnum.SUCCESS);
+  }
```
We test the `creation` and `getByKey` here, why not the search as well? What is our testing concept/purpose/goal here?

koevskinikola replied:
We test the search API, not `creation`. The test concept for the `AuditLog[Rest\GRPC]ClientIT` tests is to cover the Java Client and REST API endpoints.

eppdot commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AuditLogAuthorizationsIT.java:
```diff
@@ -0,0 +1,275 @@
+      @Authenticated(USER_C_USERNAME) final CamundaClient userCClient) {
+    // when
+    final var userCAuditLogs =
+        userCClient
```
🔧 how about naming this something like `userTaskUserClient` or so? A/B/C is pretty generic 😄

eppdot commented on service/src/main/java/io/camunda/service/AuditLogServices.java:
```diff
@@ -71,9 +73,14 @@ public AuditLogEntity getAuditLog(final String auditLogKey) {
                 .getAuditLog(auditLogKey));
   }
+
+  private AuthorizationCondition getRequiredAuthorizations() {
```
🔧  can this be static and/or a constant?

eppdot reviewed (dismissed):
I have still a few open questions 😄 maybe it makes sense to jump on a quick sync call. Probably, I am just not getting it.

koevskinikola committed:
test: disable test on rdbms secondary storage

Disable the audit log acceptance tests until the
audit log search layer supports RDBMS storage
querying.

koevskinikola commented on qa/acceptance-tests/src/test/java/io/camunda/it/auditlog/AuditLogAuthorizationsIT.java:
```diff
@@ -0,0 +1,275 @@
+    assertThat(userDAuditLogs.items().size()).isOne();
+    assertThat(
+            userDAuditLogs.items().stream()
+                .allMatch(log -> PROCESS_ID.equals(log.getProcessDefinitionId())))
```
It didn't make sense to parameterize any of the tests. It would have created more work and boilerplate code that is already reduced by the JUnit 5 extension.

konstantinos-mylosis reviewed (approved):
Ran through comments
Already reviewed
Also LGTM
Nice!
